### PR TITLE
Implement battery drain and recharge for slotcars

### PR DESCRIPTION
**Pull request by mrushyendra**
Computes battery drain for each slotcar in each time step and publishes the states of charge on the `/robot_state` topic. Battery drain is computed as a sum of static drain (constant over time) and dynamic drain (due to robot's mass and trajectory), similar to the `simpleDevicePowerSink` and `simpleMotionPowerSink` in `rmf_core`. Also computes recharge if robot is stationary at a charger waypoint.

Adds Ignition and Gazebo classic GUI plugins to enable and disable charge, discharge and instant charge. The GUI plugins publish messages using Ignition/Gazebo transport libraries which are in turn read by the respective Ignition or Gazebo slotcar plugins. I also have a version where they publish messages using ROS2, which is read by the slotcar_common executable at (https://github.com/mrushyendra/traffic_editor/tree/slotcar_charging), if that is preferred instead.

Lastly, the slotcars do not currently stop if the state of charge drops to 0. Should I change the behavior to disable movement if the battery dies?

### CODE REVIEW

**Comment by mrushyendra** [comment]
To test the Ignition GUI plugin, osrf/rmf_demos#144 is required. The `toggle_charging` plugin needs to be selected from the drop down menu on the top right.

rsz_toggle_charging_ign_gui_plugin

**Review by Yadunund — CHANGES_REQUESTED** [review]
Thanks for getting this up so quickly! I've left some comments for your consideration. But would like to highlight two glaring issues below.
1. The state of charge `_soc` is stored internally as a percentage while functions `compute_discharge` and `compute_charge` return the fraction of total capacity charged/discharged, ie `dSOC` is a fraction between [0,1]. But in your implementation, you add/subtract this fraction to/from a percentage `_soc`. As a result running the simulation for 5-6hrs only results in <.1% drain in battery. A potential fix for this is to initialize `_soc` to `1.0` along with the same when it charges up. Then we should see more reasonable battery drains. 

2. In the `slotcar_common::update()` method, does the `dist` vector contain  [dx,dy,dyaw] ? From the implementation it looks like it stores [dx,dy,dz]. If this is the case then the values of `v`, `w` and `alpha` will be incorrect. `v` should equal the norm of only the dx/dt and dy/dt components, `w = dyaw/dt` and `alpha = dw/dt`. But right now `v` is the norm dx/dt,dy/dt,dz/dt components while `w=dz/dt` (and hence `alpha` will be incorrect). Let me know if otherwise.

The toggle charger plugin looks good but the placement of the buttons can be lowered a bit as the current layout seems to interfere with the `toggle_floor` plugin button. 
image

Could we also set the `mode` field in the `RobotState` message to `MODE_CHARGING` when the battery is being charged. For now let us not render the robot immobile if the battery is fully drained.

File: building_sim_plugins/building_plugins_common/include/building_sim_common/slotcar_common.hpp [hunk 1]
```diff
@@ -184,69 +248,168 @@ class SlotcarCommon
...
+      if (waypoints->HasElement("rmf:vertex"))
+      {
+        auto waypoint = waypoints->GetElement("rmf:vertex");
+        while (waypoint)
+        {
+          if (waypoint->HasAttribute("x") && waypoint->HasAttribute("y") &&
+            waypoint->HasAttribute("level"))
+          {
+            std::string x_val, y_val, lvl_name;
+            waypoint->GetAttribute("x")->Get(x_val);
```
[the remark this answers is not in the document]
Comment (mrushyendra) [reply building_sim_plugins/building_plugins_common/include/building_sim_common/slotcar_common.hpp]: Right, I've simplified this bit.

File: building_sim_plugins/building_plugins_common/src/slotcar_common.cpp [hunk 2]
```diff
@@ -191,11 +223,49 @@ std::pair<double, double> SlotcarCommon::update(const Eigen::Isometry3d& pose,
...
+  double dt = time - _last_update_time;
   _last_update_time = time;
 
   _pose = pose;
   publish_robot_state(time);
 
+  // Update battery state of charge
+  if (_initialized_pose)
+  {
+    const Eigen::Vector3d dist = compute_dist(_old_pose, _pose);
```
[the remark this answers is not in the document]
Comment (mrushyendra) [reply building_sim_plugins/building_plugins_common/src/slotcar_common.cpp]: I somehow overlooked this! Changed it to take into account the difference in yaw angle instead.

File: building_sim_plugins/building_plugins_common/include/building_sim_common/slotcar_common.hpp [hunk 3]
```diff
@@ -166,7 +216,21 @@ class SlotcarCommon
...
-  std::string get_level_name(const double z);
+  PowerParams _params;
+  bool _enable_charge = true;
+  bool _enable_instant_charge = false;
+  bool _enable_drain = true;
+  // Used for comparing with name argument of charge_state_cb to identify button selected
+  const std::string _enable_charge_str = "_enable_charge";
+  const std::string _enable_instant_charge_str = "_enable_instant_charge";
+  const std::string _enable_drain_str = "_enable_drain";
+  double _soc = 100.0;
```
[the remark this answers is not in the document]
Comment (mrushyendra) [reply building_sim_plugins/building_plugins_common/include/building_sim_common/slotcar_common.hpp]: Thanks for catching this, changed it [0,1] instead.

**Comment by mrushyendra** [comment]
Thanks for the comments! I made the necessary changes, including changing the soc range to [0,1] and taking into account dyaw instead of dz when computing the energy loss.

File: building_sim_plugins/building_plugins_common/src/slotcar_common.cpp [hunk 4]
```diff
@@ -412,6 +515,7 @@ void SlotcarCommon::publish_state_topic(const rclcpp::Time& t)
 {
   rmf_fleet_msgs::msg::RobotState robot_state_msg;
   robot_state_msg.name = _model_name;
+  robot_state_msg.battery_percent = _soc;
```
Comment (Yadunund) [inline building_sim_plugins/building_plugins_common/src/slotcar_common.cpp:518]:
```suggestion
  robot_state_msg.battery_percent = 100 * _soc;
```

File: building_sim_plugins/building_plugins_common/src/slotcar_common.cpp [hunk 5]
```diff
@@ -454,3 +558,87 @@ void SlotcarCommon::map_cb(
...
+  // Loss through acceleration
+  const double EA = ((_params.mass * a * v) +
+    (_params.inertia * alpha * w)) * run_time;
+
+  // Loss through friction. Ignores friction due to rotation in place
+  const double EF = compute_friction_energy(_params.friction_coefficient,
+      _params.mass, v, run_time);
+
+  // Change in energy as a result of motion
+  const double dE = EA + EF;
```
Comment (Yadunund) [inline building_sim_plugins/building_plugins_common/src/slotcar_common.cpp:632]: Can we tidy up this part by renaming `dE2` to dED` for energy drain from devices and adding the same to `dE` in this step.

**Review by Yadunund — APPROVED** [review]
Thanks for addressing all the issues so quickly!

**Pull request merged**